File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        queue<pair<TreeNode*,unsigned long long int>>q;
        q.push({root,0});
        unsigned long long int ans = 0;
        while(q.size()){
            unsigned long long x = q.size();
            unsigned long long int mn;
            unsigned long long int mx;
        for(int i=0;i<x;i++){
            auto it = q.front();
            q.pop();
            if(it.first->left){
                q.push({it.first->left,(2*it.second+1)});
            }
            if(it.first->right){
                q.push({it.first->right,(2*it.second+2)});
            }
            if(i==0){
                mn=it.second;
            }
            if(i==x-1){
                mx = it.second;
            }
        }
            // cout<<mx<<" "<<mn<<endl;
            ans = max(ans,mx-mn+1);
        }
    
        return ans;
    
    }
};

File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Pair{
        TreeNode node;
        int level;
        Pair(TreeNode node, int level){
            this.node = node;
            this.level = level;
        }
    }
    public int widthOfBinaryTree(TreeNode root) {
        if(root == null) return 0;
        
        Queue<Pair> que = new ArrayDeque<>();
        
        que.add(new Pair(root, 1));
        int maxWidth = 0;
        
        while(que.size() > 0){
            int size = que.size();
            int start = que.peek().level;
            // System.out.println(start);
            while(size-- > 0){
                Pair curr = que.remove();

                if(curr.node.left != null){
                    que.add(new Pair(curr.node.left, 2*(curr.level - start)+ 1));
                }
                if(curr.node.right != null){
                    que.add(new Pair(curr.node.right, 2*(curr.level - start) + 2));
                }   
                maxWidth = Math.max(maxWidth, curr.level - start + 1);
            } 
        }
        
        return maxWidth;
    }
}